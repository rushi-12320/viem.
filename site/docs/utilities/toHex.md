---
head:
  - - meta
    - property: og:title
      content: toHex
  - - meta
    - name: description
      content: Encodes a string, number, boolean or byte array to a hex value value.
  - - meta
    - property: og:description
      content: Encodes a string, number, boolean or byte array to a hex value value.

---

# toHex

Encodes a string, number, boolean or byte array to a hex value value.

## Import

```ts
import { toHex } from 'viem'
```

## Usage

```ts
import { toHex } from 'viem'

toHex(420)
// "0x1a4"

toHex('Hello world')
// "0x48656c6c6f20776f726c642e"

toHex(
  new Uint8Array([72, 101, 108, 108, 111, 32, 87, 111, 114, 108, 100, 33])
)
// "0x48656c6c6f20576f726c6421"

toHex(true)
// "0x1"
```

## Returns

[`Hex`](/docs/glossary/types#hex)

The hex value.

## Parameters

### Value

- **Type:** `string | number | bigint | ByteArray`

The value to hex encode.

## Shortcut Functions

### numberToHex

- **Type:** `number | bigint`

Encodes a number value to a hex value.

```ts
import { numberToHex } from 'viem'

numberToHex(420)
// "0x1a4"

numberToHex(4206942069420694206942069420694206942069n)
// "0xc5cf39211876fb5e5884327fa56fc0b75"
```

### stringToHex

- **Type:** `string`

Encodes a UTF-8 string value to a hex value.

```ts
import { stringToHex } from 'viem'

stringToHex('Hello World!')
// "0x48656c6c6f20576f726c6421"
```

### bytesToHex

- **Type:** `ByteArray`

Encodes a byte array to a hex value.

```ts
import { stringToHex } from 'viem'

bytesToHex(
  new Uint8Array([72, 101, 108, 108, 111, 32, 87, 111, 114, 108, 100, 33]),
)
// "0x48656c6c6f20576f726c6421"
```

### boolToHex

- **Type:** `boolean`

Encodes a boolean to a hex value.

```ts
import { stringToHex } from 'viem'

bytesToHex(true)
// "0x1"
```