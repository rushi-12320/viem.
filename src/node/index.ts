export {
  type IpcTransport,
  type IpcTransportConfig,
  type IpcTransportErrorType,
  ipc,
} from '../clients/transports/ipc.js'

export { mainnetTrustedSetup } from './trustedSetups.js'

export {
  type IpcRpcClient,
  getIpcRpcClient,
} from '../utils/rpc/ipc.js'
