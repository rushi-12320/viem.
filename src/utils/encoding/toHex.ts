import { pad } from '../data'
import type { ByteArray, Hex } from '../../types'
import { IntegerOutOfRangeError } from '../../errors'

const hexes = Array.from({ length: 256 }, (_v, i) =>
  i.toString(16).padStart(2, '0'),
)

/**
 * @description Encodes a boolean into a hex string
 */
export function boolToHex(value: boolean): Hex {
  return `0x${Number(value)}`
}

/**
 * @description Encodes a bytes array into a hex string
 */
export function bytesToHex(value: ByteArray): Hex {
  let hex = ''
  for (let i = 0; i < value.length; i++) {
    hex += hexes[value[i]]
  }
  return `0x${hex}`
}

/**
 * @description Encodes a string, number, bigint, or ByteArray into a hex string
 */
export function toHex(
  value: string | number | bigint | boolean | ByteArray,
): Hex {
  if (typeof value === 'number' || typeof value === 'bigint')
    return numberToHex(value)
  if (typeof value === 'string') {
    return stringToHex(value)
  }
  if (typeof value === 'boolean') return boolToHex(value)
  return bytesToHex(value)
}

export type NumberToHexOpts =
  | {
      // Whether or not the number of a signed representation.
      signed?: boolean
      // The size of the output hex (in bytes).
      size: number
    }
  | {
      signed?: never
      size?: never
    }

/**
 * @description Encodes a number or bigint into a hex string
 */
export function numberToHex(
  value_: number | bigint,
  opts: NumberToHexOpts = {},
): Hex {
  const { signed, size } = opts

  let value = BigInt(value_)

  let maxValue
  if (size) {
    if (signed) maxValue = (1n << (BigInt(size) * 8n - 1n)) - 1n
    else maxValue = 2n ** (BigInt(size) * 8n) - 1n
  } else if (typeof value_ === 'number') {
    maxValue = BigInt(Number.MAX_SAFE_INTEGER)
  }

  const minValue = typeof maxValue === 'bigint' && signed ? -maxValue - 1n : 0

  if ((maxValue && value > maxValue) || value < minValue) {
    const suffix = typeof value_ === 'bigint' ? 'n' : ''
    throw new IntegerOutOfRangeError({
      max: maxValue ? `${maxValue}${suffix}` : undefined,
      min: `${minValue}${suffix}`,
      signed,
      size,
      value: `${value_}${suffix}`,
    })
  }

  const hex = `0x${(signed && value < 0
    ? (1n << BigInt(size * 8)) + BigInt(value)
    : value
  ).toString(16)}` as Hex
  if (size) return pad(hex, { size }) as Hex
  return hex
}

const encoder = new TextEncoder()

/**
 * @description Encodes a UTF-8 string into a hex string
 */
export function stringToHex(value_: string): Hex {
  const value = encoder.encode(value_)
  return toHex(value)
}
