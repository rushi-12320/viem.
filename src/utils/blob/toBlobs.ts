import {
  bytesPerBlob,
  fieldElementsPerBlob,
  maxBytesPerTransaction,
} from '../../constants/blob.js'
import { BlobSizeTooLarge, EmptyBlobError } from '../../errors/blob.js'
import type { ByteArray, Hex } from '../../types/misc.js'
import { size } from '../data/size.js'
import { hexToBytes } from '../encoding/toBytes.js'
import { bytesToHex } from '../encoding/toHex.js'

type To = 'hex' | 'bytes'

export type ToBlobsReturnType<to extends To> =
  | (to extends 'bytes' ? ByteArray[] : never)
  | (to extends 'hex' ? Hex[] : never)

export function toBlobs<
  const data extends Hex | ByteArray,
  to extends To =
    | (data extends Hex ? 'hex' : never)
    | (data extends ByteArray ? 'bytes' : never),
>(
  data_: data | Hex | ByteArray,
  to_?: to | To | undefined,
): ToBlobsReturnType<to> {
  const to = to_ ?? (typeof data_ === 'string' ? 'hex' : 'bytes')
  const data = typeof data_ === 'string' ? hexToBytes(data_) : data_

  const size_ = size(data)
  if (!size_) throw new EmptyBlobError()
  if (size_ > maxBytesPerTransaction)
    throw new BlobSizeTooLarge({ maxSize: maxBytesPerTransaction, size: size_ })

  const length = Math.ceil(size_ / bytesPerBlob)
  const paddedData = pad(data as any, length)

  const blobs: Uint8Array[] = []
  for (let i = 0; i < length; i++) {
    const chunk = paddedData.subarray(i * bytesPerBlob, (i + 1) * bytesPerBlob)
    const blob = toBlob(chunk)
    blobs.push(blob)
  }

  return (
    to === 'bytes' ? blobs : blobs.map((x) => bytesToHex(x))
  ) as ToBlobsReturnType<to>
}

function pad(data_: ByteArray, length: number): ByteArray {
  const data = new Uint8Array(length * bytesPerBlob).fill(0)
  data.set(data_)
  data[data_.byteLength] = 0x80
  return data
}

function toBlob(data: Uint8Array): Uint8Array {
  const blob = new Uint8Array(bytesPerBlob)
  for (let i = 0; i < fieldElementsPerBlob; i++) {
    const chunk = new Uint8Array(32)
    chunk.set(data.subarray(i * 31, (i + 1) * 31), 0)
    blob.set(chunk, i * 32)
  }

  return blob
}
