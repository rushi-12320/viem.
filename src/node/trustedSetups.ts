import { resolve } from 'node:path'

export const mainnetTrustedSetup = resolve(
  __dirname,
  '../trusted-setups/mainnet.txt',
)
