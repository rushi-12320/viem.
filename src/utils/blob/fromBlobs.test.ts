import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { expect, test } from 'vitest'
import { stringToHex } from '../encoding/toHex.js'
import { fromBlobs } from './fromBlobs.js'
import { toBlobs } from './toBlobs.js'

test('default', () => {
  const data = stringToHex('we are all gonna make it'.repeat(5))
  const blobs = toBlobs({
    data,
  })
  expect(
    fromBlobs({
      blobs,
    }),
  ).toEqual(data)
})

test('large', () => {
  const data = stringToHex(
    readFileSync(
      resolve(__dirname, '../../../test/kzg/lorem-ipsum.txt'),
      'utf-8',
    ),
  )
  const blobs = toBlobs({
    data,
  })
  expect(
    fromBlobs({
      blobs,
    }),
  ).toEqual(data)
})
