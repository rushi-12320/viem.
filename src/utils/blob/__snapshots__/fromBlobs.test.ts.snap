// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`default 1`] = `
Uint8Array [
  119,
  101,
  32,
  97,
  114,
  101,
  32,
  97,
  108,
  108,
  32,
  103,
  111,
  110,
  110,
  97,
  32,
  109,
  97,
  107,
  101,
  32,
  105,
  116,
  119,
  101,
  32,
  97,
  114,
  101,
  32,
  97,
  108,
  108,
  32,
  103,
  111,
  110,
  110,
  97,
  32,
  109,
  97,
  107,
  101,
  32,
  105,
  116,
  119,
  101,
  32,
  97,
  114,
  101,
  32,
  97,
  108,
  108,
  32,
  103,
  111,
  110,
  110,
  97,
  32,
  109,
  97,
  107,
  101,
  32,
  105,
  116,
  119,
  101,
  32,
  97,
  114,
  101,
  32,
  97,
  108,
  108,
  32,
  103,
  111,
  110,
  110,
  97,
  32,
  109,
  97,
  107,
  101,
  32,
  105,
  116,
  119,
  101,
  32,
  97,
  114,
  101,
  32,
  97,
  108,
  108,
  32,
  103,
  111,
  110,
  110,
  97,
  32,
  109,
  97,
  107,
  101,
  32,
  105,
]
`;
